types/match.go

```diff
@@ -19,16 +19,18 @@ import (
 	"regexp"
 	"sort"
 
+	"github.com/prometheus/alertmanager/pkg/labels"
 	"github.com/prometheus/common/model"
 )
 
 // Matcher defines a matching rule for the value of a given label.
 type Matcher struct {
-	Name    string `json:"name"`
-	Value   string `json:"value"`
-	IsRegex bool   `json:"isRegex"`
-
-	regex *regexp.Regexp
+	Type            labels.MatchType
+	Name            string `json:"name"`
+	Value           string `json:"value"`
+	IsRegex         bool   `json:"isRegex"`
+	NewRouteMatcher bool
```

Comment:
I tried implementing (3) but ran into some issues. It would be really helpful if you could help me out with them.

I created a new type in config.go which unmarshals the yaml and uses labels.matchers. However, I am not able to understand how that translates to not modifying any code in[ match.go](https://github.com/prometheus/alertmanager/blob/master/types/match.go), since route accepts Matchers of type, [types.Matchers] (https://github.com/prometheus/alertmanager/blob/master/dispatch/route.go#L91) , we would need a new type.matcher which means creating a new constructor anyways. 

Apart from this, checking error while unmarshalling sounds good but I am confused when UnmarshalYAML is called. I am guessing it is called automatically but am not able to understand the flow of the code. 

cc @beorn7

Reply:
Yes, we would need to switch over the routing to use `labels.Matchers`. With #2441 included, this should keep all existing configs in working state. `types.Matcher` would only be used by silences (and possibly get kicked out there eventually, but let's not think about that now).

I guess my cunning plan might collide with some devil in the detail. I could play with it now, but how about some pair programming tomorrow, if the pager permits? Let's sync up once I'm awake…

Reply:
Yep sure. Sent you an invite !